### PR TITLE
Filter multiauth authentication sources from SP using AuthnContextClassRef

### PR DESCRIPTION
**Pull request by Lasering**
This PR introduces the necessary changes to allow an SP to choose an authentication source of multiauth. It works even if there are redirects during the authentication process because it uses SAML2 standard AuthnContextClassRef. See #912 for more context.

### CODE REVIEW

File: modules/multiauth/docs/multiauth.md
Comment (tvdijen) [inline modules/multiauth/docs/multiauth.md:82]: You're missing `Authn`

Comment (Lasering) [reply modules/multiauth/docs/multiauth.md]: Initially I implemented it with AuthnContextClassRef. But it seemed weird to reference Authn in the multiauth config. So I changed it to ContextClassRef. I've reverted back to AuthnContextClassRef.

**Comment by Lasering** [comment]
I've set AuthnContextClassRef in multiauth as optional. If the SP defines an AuthnContextClassRef but multiauth config does not, should the NoAuthnContext be returned?

How can I return a `SAML2\XML\samlp\StatusCode(SAML2\Constants::STATUS_NO_AUTHN_CONTEXT)` from inside the authenticate function (assuming that is the right thing to return)?

**Comment by tvdijen** [comment]
Hmm, this is quite tricky.. I think we need a defaultAuthnContext outside the sources-array..
If the requested context doesn't match a specific source or the default one, it should definitely return an error.
I think I'd like a second opinion by @jaimeperez to make sure we don't miss anything..

To return the proper error, I think it's enough to just `throw \SimpleSAML\Module\saml\Error\NoAuthnContext(\SAML2\Constants::STATUS_RESPONDER);`..

**Comment by jaimeperez** [comment]
Thanks a lot for your contribution @Lasering!

This one is indeed tricky. I think we might need to explicitly enable this in the `multiauth` configuration, unless as Tim was suggesting, we provide a default that can be applied to those sources missing it in the configuration, and then use that as an indication that the functionality is enabled. But we definitely need a clear way to enable/disable this.

Now, I see at least a couple other issues that I think we need to address:

* Should we have a way to provide, by configuration, mappings for known `AuthnContextClassRef`s? In my head, a defined auth source could match several class refs, as those are just references to an agreed set of conditions and requirements.
* We need to be extra careful with the comparison algorithm, and decide what we support. Anything other than `exact` is quite a nightmare to implement after my own experience, but it might limit a lot the usefulness of the implementation (e.g. LoA2 with comparison `minimum`).

**Comment by tvdijen** [comment]
The default comparison should be `exact` because that's what SAML2INT mandates. Others could be optional.

**Comment by Lasering** [comment]
If the requested context class refs only match the default value for AuthnContextClassRef which auth method should be shown? The way I see it giving the error when 0 auth methods match is already the best approach.

1. If multiauth does not define any class refs and the SP requests some refs => the error NoAuthnContext will be thrown, which is good because it signals the IdP maintainer of the problem (we could even suggest in the error message to add class refs to the multiauth config).
2. If multiauth defines refs and the SP does not => all auth methods will be shown. Which intuitively makes sense.
3. If both define refs => the list of auth methods will be filtered. If 0 methods remain after filtering the error will be thrown.

> * Should we have a way to provide, by configuration, mappings for known `AuthnContextClassRef`s? In my head, a defined auth source could match several class refs, as those are just references to an agreed set of conditions and requirements.

This is already solved. The AuthnContextClassRef of a source in multiauth can be a string or an array of strings. This way you can specify that an authentication source corresponds to multiple class refs.

> * We need to be extra careful with the comparison algorithm, and decide what we support. Anything other than `exact` is quite a nightmare to implement after my own experience, but it might limit a lot the usefulness of the implementation (e.g. LoA2 with comparison `minimum`).

Before creating the PR I went looking for that logic, since I assumed it should exist in either simplesamlphp or in the saml2 library, but I could not find it. The spec is also unhelpful since:

> If Comparison is set to _______ then the resulting authentication context in the authentication statement MUST _______.
> 
> - `exact` or omitted - be the exact match of at least one of the authentication contexts specified.
> - `minimum` - be at least as strong (as deemed by the responder) as one of the authentication contexts specified.
> - `better` - be stronger (as deemed by the responder) than any one of the authentication contexts specified.
> - `maximum` - be as strong as possible (as deemed by the responder) without exceeding the strength of at least one of the authentication contexts specified.

1. The choice of what is at least as strong/stronger/as strong is left to the responder (the IdP in this scenario).
2. Currently the field AuthnContextClassRef in simplesamlphp is an unrestricted string.

The IdP would have to somehow assign a strongness factor (for the lack of a better description) to each class ref, and then perform the comparisons based on that factor. These assignments would need to be a configuration since 1) it is very unlikely that everyone will agree on the same assignments 2) that is the only way of having a change to deal with the unrestricted string.

We could simply declare the only supported implementation for the comparison is `exact`. The use cases supported by the current implementation are already useful enough.

**Comment by tvdijen** [comment]
> Before creating the PR I went looking for that logic, since I assumed it should exist in either simplesamlphp or in the saml2 library, but I could not find it.

The spec does say something in this regard on the RequestedAuthnContext:
_The set of supplied references MUST be evaluated as an ordered set, where the first element is the most preferred authentication context class or declaration._
So we start at the top and see if we can satisfy the requirement.. If not, move our way down the list and eventually if nothing works, the NoAuthnContext-exception must be thrown.

If we continue this practise or ordering the LOA's to the IDP (make sure we add it to the documentation), We should be able to handle minimum/maximum/better..

I think first we need to determine a list of possible LOA's that fit the requirement.. For example, we have two sources:
A: [LOA2, LOA1]
B: [LOA3, LOA2]
C: [LOA4, LOA3]
D: [LOA4]

Now the SP requires 'better' than LOA1.. We loop over the sources to find that A can meet this requirement.. Now we should do another loop over the sources to figure out that B can also satisfy LOA2 (or even LOA4 or LOA3, because they're higher in the ordered list). Now that we have determined that we could meet the requirement using [LOA4, LOA3, LOA2], we should be able to figure out that source D is also suitable.

I guess if we systematically peel this off for minimum/maximum/better/exact we should be able to make this work.

**Comment by jaimeperez** [comment]
> 1. If multiauth does not define any class refs and the SP requests some refs => the error NoAuthnContext will be thrown, which is good because it signals the IdP maintainer of the problem (we could even suggest in the error message to add class refs to the multiauth config).

Not sure I agree with this one, as that would break existing deployments. I think this needs to be enabled explicitly in the IdP, in a way that doesn't enforce it for existing configurations.

> 2. If multiauth defines refs and the SP does not => all auth methods will be shown. Which intuitively makes sense.

Agree.

> 3. If both define refs => the list of auth methods will be filtered. If 0 methods remain after filtering the error will be thrown.

Agree as well.

> > * Should we have a way to provide, by configuration, mappings for known `AuthnContextClassRef`s? In my head, a defined auth source could match several class refs, as those are just references to an agreed set of conditions and requirements.
> 
> This is already solved. The AuthnContextClassRef of a source in multiauth can be a string or an array of strings. This way you can specify that an authentication source corresponds to multiple class refs.

Awesome! Sorry that I replied without actually having a look at the code 😅 

> The IdP would have to somehow assign a strongness factor (for the lack of a better description) to each class ref, and then perform the comparisons based on that factor. These assignments would need to be a configuration since 1) it is very unlikely that everyone will agree on the same assignments 2) that is the only way of having a change to deal with the unrestricted string.

This can be very difficult to assess in practice, so I would agree with Tim's initial suggestion and your comment to essentially support `exact`. I'd say it would be nice to support `minimum` as well, but I wouldn't focus on it now. As long as we can define a set of different class refs for each auth source, so that we can compare each of them with whatever we receive from the SP, `exact` is fairly simple and should work in most deployments.

Question: would the _default_ class ref apply also to those auth sources explicitly specifying any class ref? I'd say no, but would like your opinions as well.

**Comment by tvdijen** [comment]
That's a no on the question.. Also something that we need to explicitly document

**Comment by Lasering** [comment]
> The set of supplied references MUST be evaluated as an ordered set, where the first element is the most preferred authentication context class or declaration.

In my interpretation `most preferred` has a different meaning from `at least as strong/stronger/as strong as possible`.
I don't follow your logic on LOA (does that mean Level of Assurance?). How it is related?

I've been assuming the Comparison attribute can only be unset (the same as exact) or set to **one** of the possible values. That together with the `most preferred` approach makes the implementation easy.

> Not sure I agree with this one, as that would break existing deployments. I think this needs to be enabled explicitly in the IdP, in a way that doesn't enforce it for existing configurations.

I'll need to make a test to confirm this but the existing implementation (without this PR) is incorrect, if the SP is explicitly asking for ContextClassRefs the code should return the NoAuthnContext (as per the spec), and not allow the user to pick auth methods. Existing deployments will break but they were working "outside" of the spec, so in a sense they were wrong from the start. Propagating this wrongness seems the worst option of the two.

Instead of having a flag enabling this functionality in the IdP, I suggest having a flag to enable working "outside" of the spec. Something like:
```php
/**
 * Whether the user will still be allowed to choose between the different authentication methods if an
 * SP requests AuthnContextClassRefs but the multiauth configuration does not set AuthnContextClassRefs
 * for any of the methods.
 * Setting this true, will make this IdP non-compliant with the [SAML2 Protocol](https://docs.oasis-open.org/security/saml/v2.0/saml-core-2.0-os.pdf) point 3.3.2.2.1
 */
'multiauth.always-list-auth-methods' => false,
```

**Comment by Lasering** [comment]
Changed implementation to invoke delegateAuthentication directly when only one authentication source remains after filtering using the ClassRefs. Save the correct list of sources in the state.

**Review by tvdijen — CHANGES_REQUESTED** [review]
Unfortunately the tests are failing.. Please fix the code to comply with them, or adjust them accordingly to the change.

**Comment by Lasering** [comment]
@tvdijen did you implement new tests? One of the failing tests is testing the preselect which I did not touch. And the tests were passing before.

**Comment by tvdijen** [comment]
I think I may have flipped the order during the rebase.. It's good again

**Review by tvdijen — APPROVED** [review]

**Comment by dbollaer** [comment]
@tvdijen in which version of simplesamlphp will this feature be released?
@Lasering thank you!

**Comment by tvdijen** [comment]
Either 2.0 or 1.20, which ever comes first..

**Pull request merged**
